[<< back to main index](../README.md) 

# Lab : Spark Dataframes


### Overview
First look at Spark Dataframes

### Depends On 
None

### Run time
20-30 mins


In [1]:
# initialize Spark Session
import os
import sys
top_dir = os.path.abspath(os.path.join(os.getcwd(), "../"))
if top_dir not in sys.path:
    sys.path.append(top_dir)

from init_spark import init_spark
spark = init_spark()
spark

Initializing Spark...
Spark found in :  /Users/sujee/spark
Spark config:
	 spark.app.name=TestApp
	spark.master=local[*]
	executor.memory=2g
	spark.sql.warehouse.dir=/var/folders/lp/qm_skljd2hl4xtps5vw0tdgm0000gn/T/tmp6vmgwg88
	some_property=some_value
Spark UI running on port 4040



## STEP 1: Load Clickstream data

In [2]:
clickstreamDF = spark.read.json("/data/click-stream/clickstream.json")
print(clickstreamDF)

DataFrame[action: string, campaign: string, cost: bigint, domain: string, ip: string, session: string, timestamp: bigint, user: string]


**==> Monitor Spark shell UI on port 4040**  
You may see something like this:

**==> Q : Why is Spark not lazy loading the JASON files? **

<img src="../assets/images/5.1a.png" style="border: 5px solid grey; max-width:100%;" />




## STEP 2 : Inspecting The Dataframe

**==> Print the schema of data frame**     


In [3]:
# TODO: Print the schema
# Hint : `printSchema()`

clickstreamDF.printSchema()

root
 |-- action: string (nullable = true)
 |-- campaign: string (nullable = true)
 |-- cost: long (nullable = true)
 |-- domain: string (nullable = true)
 |-- ip: string (nullable = true)
 |-- session: string (nullable = true)
 |-- timestamp: long (nullable = true)
 |-- user: string (nullable = true)



Your output may look like this:

```
    root
     |-- action: string (nullable = true)
     |-- campaign: string (nullable = true)
     |-- cost: long (nullable = true)
     |-- domain: string (nullable = true)
     |-- ip: string (nullable = true)
     |-- session: string (nullable = true)
     |-- timestamp: long (nullable = true)
     |-- user: string (nullable = true)
```

**==> Print / Dump the data contained within dataframe**  

In [4]:
# TODO: Print/Dump the data
# Hint : show()
clickstreamDF.show()

+-------+-----------+----+-----------------+----+----------+-------------+------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
+-------+-----------+----+-----------------+----+----------+-------------+------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
|blocked|campaign_12|   5|     facebook.com|ip_3|session_96|1420070400864|user_5|
|clicked| campaign_3|  54|sf.craigslist.org|ip_9|session_61|1420070401728|user_8|
|blocked|campaign_18| 110|    wikipedia.org|ip_5|session_55|1420070402592|user_6|
|clicked| campaign_6|  15|comedycentral.com|ip_9|session_49|1420070403456|user_4|
|blocked| campaign_9| 139|          cnn.com|ip_8|session_13|1420070404320|user_7|
|blocked| campaign_4| 171|   funnyordie.com|ip_1|session_92|1420070405184|user_9|
|blocked|campaign_17|  20|       amazon.com|ip_4|session_13|1420070406048|user_1|
|blocked|campaign_20|  78|        zynga.com|ip_5|session_36|1420070406912|user_3|
|blocked|campaig

Your output may look like this:

```
+-------+-----------+----+-----------------+----+----------+-------------+------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
+-------+-----------+----+-----------------+----+----------+-------------+------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
|blocked|campaign_12|   5|     facebook.com|ip_3|session_96|1420070400864|user_5|
|clicked| campaign_3|  54|sf.craigslist.org|ip_9|session_61|1420070401728|user_8|
...

```

**==> Explore methods available in Dataframe**  
1) Here is the Dataframe API : 
[Scala](http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.sql.DataFrame)  /
[Java](http://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/DataFrame.html) / 
[Python](http://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=dataframe#pyspark.sql.DataFrame)



## STEP 3 : Querying Dataframe

**==> Show only click logs where the cost > 100**

In [5]:
clickstreamDF.filter("cost > 100").show()

+-------+-----------+----+-----------------+----+----------+-------------+------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
+-------+-----------+----+-----------------+----+----------+-------------+------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
|blocked|campaign_18| 110|    wikipedia.org|ip_5|session_55|1420070402592|user_6|
|blocked| campaign_9| 139|          cnn.com|ip_8|session_13|1420070404320|user_7|
|blocked| campaign_4| 171|   funnyordie.com|ip_1|session_92|1420070405184|user_9|
|blocked|campaign_19| 147|      nytimes.com|ip_2|session_65|1420070407776|user_6|
|clicked|campaign_16| 141|    wikipedia.org|ip_1|session_57|1420070409504|user_1|
|blocked| campaign_9| 153|       google.com|ip_3|session_22|1420070413824|user_9|
|blocked| campaign_8| 140|comedycentral.com|ip_8| session_4|1420070414688|user_2|
| viewed|campaign_20| 133|       google.com|ip_9|session_69|1420070416416|user_7|
+-------+-------

In [6]:
#Same thing
clickstreamDF.filter(clickstreamDF['cost'] > 100).show()


+-------+-----------+----+-----------------+----+----------+-------------+------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
+-------+-----------+----+-----------------+----+----------+-------------+------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
|blocked|campaign_18| 110|    wikipedia.org|ip_5|session_55|1420070402592|user_6|
|blocked| campaign_9| 139|          cnn.com|ip_8|session_13|1420070404320|user_7|
|blocked| campaign_4| 171|   funnyordie.com|ip_1|session_92|1420070405184|user_9|
|blocked|campaign_19| 147|      nytimes.com|ip_2|session_65|1420070407776|user_6|
|clicked|campaign_16| 141|    wikipedia.org|ip_1|session_57|1420070409504|user_1|
|blocked| campaign_9| 153|       google.com|ip_3|session_22|1420070413824|user_9|
|blocked| campaign_8| 140|comedycentral.com|ip_8| session_4|1420070414688|user_2|
| viewed|campaign_20| 133|       google.com|ip_9|session_69|1420070416416|user_7|
+-------+-------

Sample output

```
    +-------+-----------+----+-----------------+----+----------+-------------+------+
    | action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
    +-------+-----------+----+-----------------+----+----------+-------------+------+
    |clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
    |blocked|campaign_18| 110|    wikipedia.org|ip_5|session_55|1420070402592|user_6|
    |blocked| campaign_9| 139|          cnn.com|ip_8|session_13|1420070404320|user_7|
```

**==> Show the logs where action = clicked**  

In [7]:
# TODO: show the logs where action = clicked
# Hint : clickstreamDF.filter("action == '???'")
clickstreamDF.filter ("action == 'clicked'")

DataFrame[action: string, campaign: string, cost: bigint, domain: string, ip: string, session: string, timestamp: bigint, user: string]

In [8]:
## TODO Another approach
clickstreamDF.filter(clickstreamDF['action'] == 'clicked').show()

+-------+-----------+----+-----------------+----+----------+-------------+------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|
+-------+-----------+----+-----------------+----+----------+-------------+------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|
|clicked| campaign_3|  54|sf.craigslist.org|ip_9|session_61|1420070401728|user_8|
|clicked| campaign_6|  15|comedycentral.com|ip_9|session_49|1420070403456|user_4|
|clicked|campaign_16| 141|    wikipedia.org|ip_1|session_57|1420070409504|user_1|
|clicked| campaign_5|  74|       sfgate.com|ip_6|session_27|1420070412096|user_3|
+-------+-----------+----+-----------------+----+----------+-------------+------+



## Step 5 : Count the number of visits from each domain

Hint : `g = clickstreamDF.groupBy("column_name")`  
* Inspect methods on `g`  (use tab completion)  
* Then do `count` on `g`  
* Then do a show  
* So `clickstreamDF.groupby().count.show`  

In [9]:
## TODO: 
g = clickstreamDF.groupBy("domain")

## TODO: do the count
# g.???()

## And finally show
g.count().show()

+-----------------+-----+
|           domain|count|
+-----------------+-----+
|      nytimes.com|    1|
|      youtube.com|    2|
|        zynga.com|    1|
|       google.com|    2|
|        yahoo.com|    1|
|     facebook.com|    1|
|          cnn.com|    1|
|    wikipedia.org|    3|
|       sfgate.com|    1|
|       amazon.com|    2|
|   funnyordie.com|    1|
|sf.craigslist.org|    2|
|comedycentral.com|    2|
+-----------------+-----+



## STEP 6 : Joining Dataframes

Let's load another data set `domain info`  
The data is in   `/data/click-stream/domain-info.json`  
The data looks like this:

```json
    {"domain":"amazon.com","category":"SHOPPING"}
    {"domain":"bbc.co.uk","category":"NEWS"}
    {"domain":"facebook.com","category":"SOCIAL"}
    ...
```

**==> Load the dataframe**

In [10]:
domainsDF = spark.read.json("/data/click-stream/domain-info.json")
print(domainsDF)


DataFrame[category: string, domain: string]


In [11]:
## TODO display domains data
## Hint : show()
domainsDF.show()

+-----------+-----------------+
|   category|           domain|
+-----------+-----------------+
|   SHOPPING|       amazon.com|
|       NEWS|        bbc.co.uk|
|     SOCIAL|     facebook.com|
|      PHOTO|       flickr.com|
|       NEWS|      foxnews.com|
|     SEARCH|       google.com|
|       NEWS|          npr.org|
|       NEWS|      nytimes.com|
|CLASSIFIEDS|sf.craigslist.org|
|       NEWS|       sfgate.com|
|     SOCIAL|      twitter.com|
|       NEWS|     usatoday.com|
|  REFERENCE|    wikipedia.org|
|     SEARCH|        yahoo.com|
|      VIDEO|      youtube.com|
+-----------+-----------------+



**==> Join both dataframes**

In [12]:

joined = clickstreamDF.join(domainsDF,  clickstreamDF["domain"] == domainsDF["domain"])
    
#see the results
joined.show()

+-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
| action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|   category|           domain|
+-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
|clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|      VIDEO|      youtube.com|
|blocked|campaign_12|   5|     facebook.com|ip_3|session_96|1420070400864|user_5|     SOCIAL|     facebook.com|
|clicked| campaign_3|  54|sf.craigslist.org|ip_9|session_61|1420070401728|user_8|CLASSIFIEDS|sf.craigslist.org|
|blocked|campaign_18| 110|    wikipedia.org|ip_5|session_55|1420070402592|user_6|  REFERENCE|    wikipedia.org|
|blocked|campaign_17|  20|       amazon.com|ip_4|session_13|1420070406048|user_1|   SHOPPING|       amazon.com|
|blocked|campaign_19| 147|      nytimes.com|ip_2|session_65|1420070407776|user_6|       NEWS|      nytim

**==> Inspect the results, here is a sample**

```
    +-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
    | action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|   category|           domain|
    +-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
    |clicked|campaign_19| 118|      youtube.com|ip_4|session_36|1420070400000|user_9|      VIDEO|      youtube.com|
    |blocked| campaign_2|   7|      youtube.com|ip_2|session_93|1420070412960|user_1|      VIDEO|      youtube.com|
    |blocked|campaign_17|  20|       amazon.com|ip_4|session_13|1420070406048|user_1|   SHOPPING|       amazon.com|
```

**==> Note some rows are missing.  Which ones?  Why?**

**==> Do an outer join**    

In [13]:

joinedOuter = clickstreamDF.join(domainsDF,  clickstreamDF["domain"] == domainsDF["domain"], "outer")
joinedOuter.show()

+-------+-----------+----+-------------+----+----------+-------------+------+---------+-------------+
| action|   campaign|cost|       domain|  ip|   session|    timestamp|  user| category|       domain|
+-------+-----------+----+-------------+----+----------+-------------+------+---------+-------------+
|blocked|campaign_19| 147|  nytimes.com|ip_2|session_65|1420070407776|user_6|     NEWS|  nytimes.com|
|clicked|campaign_19| 118|  youtube.com|ip_4|session_36|1420070400000|user_9|    VIDEO|  youtube.com|
|blocked| campaign_2|   7|  youtube.com|ip_2|session_93|1420070412960|user_1|    VIDEO|  youtube.com|
|blocked|campaign_20|  78|    zynga.com|ip_5|session_36|1420070406912|user_3|     null|         null|
|blocked| campaign_9| 153|   google.com|ip_3|session_22|1420070413824|user_9|   SEARCH|   google.com|
| viewed|campaign_20| 133|   google.com|ip_9|session_69|1420070416416|user_7|   SEARCH|   google.com|
|   null|       null|null|         null|null|      null|         null|  null|     

**==> Inspect the output, might look like this**  
**==> Can you explain the null values?**

Output:

```
    +-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
    | action|   campaign|cost|           domain|  ip|   session|    timestamp|  user|   category|           domain|
    +-------+-----------+----+-----------------+----+----------+-------------+------+-----------+-----------------+
    |blocked| campaign_9| 139|          cnn.com|ip_8|session_13|1420070404320|user_7|       null|             null|
    |   null|       null|null|             null|null|      null|         null|  null|     SOCIAL|      twitter.com|
    |clicked| campaign_6|  15|comedycentral.com|ip_9|session_49|1420070403456|user_4|       null|             null|
```

## Step 7: Understanding Query Execution
We will use **explain** keyword

In [14]:
clickstreamDF.filter("cost > 100").explain(extended=True)

== Parsed Logical Plan ==
'Filter ('cost > 100)
+- AnalysisBarrier
      +- Relation[action#6,campaign#7,cost#8L,domain#9,ip#10,session#11,timestamp#12L,user#13] json

== Analyzed Logical Plan ==
action: string, campaign: string, cost: bigint, domain: string, ip: string, session: string, timestamp: bigint, user: string
Filter (cost#8L > cast(100 as bigint))
+- Relation[action#6,campaign#7,cost#8L,domain#9,ip#10,session#11,timestamp#12L,user#13] json

== Optimized Logical Plan ==
Filter (isnotnull(cost#8L) && (cost#8L > 100))
+- Relation[action#6,campaign#7,cost#8L,domain#9,ip#10,session#11,timestamp#12L,user#13] json

== Physical Plan ==
*(1) Project [action#6, campaign#7, cost#8L, domain#9, ip#10, session#11, timestamp#12L, user#13]
+- *(1) Filter (isnotnull(cost#8L) && (cost#8L > 100))
   +- *(1) FileScan json [action#6,campaign#7,cost#8L,domain#9,ip#10,session#11,timestamp#12L,user#13] Batched: false, Format: JSON, Location: InMemoryFileIndex[file:/data/click-stream/clickstream.json

In [15]:
joined.explain(extended=True)

== Parsed Logical Plan ==
Join Inner, (domain#9 = domain#192)
:- Relation[action#6,campaign#7,cost#8L,domain#9,ip#10,session#11,timestamp#12L,user#13] json
+- Relation[category#191,domain#192] json

== Analyzed Logical Plan ==
action: string, campaign: string, cost: bigint, domain: string, ip: string, session: string, timestamp: bigint, user: string, category: string, domain: string
Join Inner, (domain#9 = domain#192)
:- Relation[action#6,campaign#7,cost#8L,domain#9,ip#10,session#11,timestamp#12L,user#13] json
+- Relation[category#191,domain#192] json

== Optimized Logical Plan ==
Join Inner, (domain#9 = domain#192)
:- Filter isnotnull(domain#9)
:  +- Relation[action#6,campaign#7,cost#8L,domain#9,ip#10,session#11,timestamp#12L,user#13] json
+- Filter isnotnull(domain#192)
   +- Relation[category#191,domain#192] json

== Physical Plan ==
*(2) BroadcastHashJoin [domain#9], [domain#192], Inner, BuildRight
:- *(2) Project [action#6, campaign#7, cost#8L, domain#9, ip#10, session#11, timesta

In [16]:
joinedOuter.explain(extended=True)

== Parsed Logical Plan ==
Join FullOuter, (domain#9 = domain#192)
:- Relation[action#6,campaign#7,cost#8L,domain#9,ip#10,session#11,timestamp#12L,user#13] json
+- Relation[category#191,domain#192] json

== Analyzed Logical Plan ==
action: string, campaign: string, cost: bigint, domain: string, ip: string, session: string, timestamp: bigint, user: string, category: string, domain: string
Join FullOuter, (domain#9 = domain#192)
:- Relation[action#6,campaign#7,cost#8L,domain#9,ip#10,session#11,timestamp#12L,user#13] json
+- Relation[category#191,domain#192] json

== Optimized Logical Plan ==
Join FullOuter, (domain#9 = domain#192)
:- Relation[action#6,campaign#7,cost#8L,domain#9,ip#10,session#11,timestamp#12L,user#13] json
+- Relation[category#191,domain#192] json

== Physical Plan ==
SortMergeJoin [domain#9], [domain#192], FullOuter
:- *(2) Sort [domain#9 ASC NULLS FIRST], false, 0
:  +- Exchange hashpartitioning(domain#9, 200)
:     +- *(1) FileScan json [action#6,campaign#7,cost#8L,dom